File: Labs/FinalExamReference/recursion/fastPowerFunction/power.m
% Octave - fast version O(log n)
function result = power_r(base, exp)
    if exp == 0
        result = 1;
    elseif mod(exp, 2) == 0
        half   = power_r(base, exp/2);
        result = half * half;
    else
        result = base * power_r(base, exp-1);
    end
end

File: Labs/FinalExamReference/recursion/fastPowerFunction/power.js
// JavaScript - fast version O(log n)
function power(base, exp) {
    if (exp === 0) return 1;
    if (exp % 2 === 0) {
        const half = power(base, exp / 2);
        return half * half;
    }
    return base * power(base, exp - 1);
}

console.log(power(2, 10));              // 1024